读取StarRocks 表时 谓词下推导致sql解析错误

sql体：
```
insert into sinkTable
select item_key,
       vehicle_id,
      item_type,item_value,modify_time
from HTWSource 
where CHAR_LENGTH(vehicle_id) < 10 ; 
```

任务启动时报错为
``` 
Caused by: java.lang.RuntimeException: Request of get query plan failed with code 500 {"exception":"The Sql is invalid","status":500}
    at com.starrocks.connector.flink.manager.StarRocksQueryPlanVisitor.getQueryPlan(StarRocksQueryPlanVisitor.java:134) ~[?:?]
    at com.starrocks.connector.flink.manager.StarRocksQueryPlanVisitor.getQueryInfo(StarRocksQueryPlanVisitor.java:64) ~[?:?]
    at com.starrocks.connector.flink.table.source.StarRocksSourceCommonFunc.getQueryInfo(StarRocksSourceCommonFunc.java:207) ~[?:?]
    at com.starrocks.connector.flink.table.source.StarRocksDynamicSourceFunction.<init>(StarRocksDynamicSourceFunction.java:83) ~[?:?]
 ```
观察日志发现sql 解析为了
```
21:06:09,066 INFO  com.starrocks.connector.flink.manager.StarRocksQueryPlanVisitor [] - query sql [select * from `bh_vehicle_iot_stream`.`htw_profile_item_info_sr` where (10)]
```

--- a/src/main/java/com/starrocks/connector/flink/table/source/StarRocksDynamicTableSource.java
+++ b/src/main/java/com/starrocks/connector/flink/table/source/StarRocksDynamicTableSource.java
@@ -14,8 +14,11 @@
 
 package com.starrocks.connector.flink.table.source;
 
-import org.apache.flink.table.api.DataTypes;
+import com.starrocks.connector.flink.table.source.struct.ColunmRichInfo;
+import com.starrocks.connector.flink.table.source.struct.PushDownHolder;
+import com.starrocks.connector.flink.table.source.struct.SelectColumn;
 
+import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.api.TableSchema;
 import org.apache.flink.table.connector.ChangelogMode;
 import org.apache.flink.table.connector.source.DynamicTableSource;
@@ -30,13 +33,10 @@ import org.apache.flink.table.expressions.ResolvedExpression;
 
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-
-import com.starrocks.connector.flink.table.source.struct.ColunmRichInfo;
-import com.starrocks.connector.flink.table.source.struct.PushDownHolder;
-import com.starrocks.connector.flink.table.source.struct.SelectColumn;
 
 public class StarRocksDynamicTableSource implements ScanTableSource, LookupTableSource, SupportsLimitPushDown, SupportsFilterPushDown, SupportsProjectionPushDown {
 
@@ -128,21 +128,27 @@ public class StarRocksDynamicTableSource implements ScanTableSource, LookupTable
     @Override
     public Result applyFilters(List<ResolvedExpression> filtersExpressions) {
         List<String> filters = new ArrayList<>();
+        List<ResolvedExpression> ac = new LinkedList<>();
+        List<ResolvedExpression> remain = new LinkedList<>();
+
         StarRocksExpressionExtractor extractor = new StarRocksExpressionExtractor();
         for (ResolvedExpression expression : filtersExpressions) {
             if (expression.getOutputDataType().equals(DataTypes.BOOLEAN()) && expression.getChildren().size() == 0) {
                 filters.add(expression.accept(extractor) + " = true");
+                ac.add(expression);
                 continue;
             }
             String str = expression.accept(extractor);
             if (str == null) {
+                remain.add(expression);
                 continue;
             }
             filters.add(str);
+            ac.add(expression);
         }
         Optional<String> filter = Optional.of(String.join(" and ", filters));
         this.pushDownHolder.setFilter(filter.get());
-        return Result.of(filtersExpressions, new ArrayList<>());
+        return Result.of(ac, remain);
     }
 
     @Override

--- a/src/main/java/com/starrocks/connector/flink/table/source/StarRocksExpressionExtractor.java
+++ b/src/main/java/com/starrocks/connector/flink/table/source/StarRocksExpressionExtractor.java
@@ -14,12 +14,6 @@
 
 package com.starrocks.connector.flink.table.source;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
-import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.expressions.CallExpression;
 import org.apache.flink.table.expressions.Expression;
 import org.apache.flink.table.expressions.ExpressionVisitor;
@@ -30,59 +24,49 @@ import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
 import org.apache.flink.table.functions.FunctionDefinition;
 import org.apache.flink.table.types.logical.LogicalTypeRoot;
 
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
 
 public class StarRocksExpressionExtractor implements ExpressionVisitor<String> {
 
-    private static final Map<FunctionDefinition, String> FUNC_TO_STR = new HashMap<>();
+    private static final Map<FunctionDefinition, Function<String[], String>> SUPPORT_FUNC = new HashMap<>();
         static {
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.EQUALS, "=");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.NOT_EQUALS, "<>");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.GREATER_THAN, ">");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.GREATER_THAN_OR_EQUAL, ">=");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.LESS_THAN, "<");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.LESS_THAN_OR_EQUAL, "<=");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.AND, "and");
-            FUNC_TO_STR.put(BuiltInFunctionDefinitions.OR, "or");
-            // FUNC_TO_STR.put(BuiltInFunctionDefinitions.IS_NULL, "is");
-            // FUNC_TO_STR.put(BuiltInFunctionDefinitions.IS_NOT_NULL, "is not");
-            // FUNC_TO_STR.put(BuiltInFunctionDefinitions.NOT, "");
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.EQUALS, args -> args[0] + " = " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.NOT_EQUALS, args -> args[0] + " <> " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.GREATER_THAN, args -> args[0] + " > " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.GREATER_THAN_OR_EQUAL, args -> args[0] + " >= " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.LESS_THAN, args -> args[0] + " < " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.LESS_THAN_OR_EQUAL, args -> args[0] + " <= " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.AND, args -> args[0] + " and " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.OR, args -> args[0] + " or " + args[1]);
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.IS_NULL, args -> args[0] + " is null");
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.IS_NOT_NULL, args -> args[0] + " is not null");
+            SUPPORT_FUNC.put(BuiltInFunctionDefinitions.NOT, args -> args[0] + " = false");
         }
 
     @Override
     public String visit(CallExpression call) {
         FunctionDefinition funcDef = call.getFunctionDefinition();
-        if (funcDef.equals(BuiltInFunctionDefinitions.LIKE)) {
-            throw new RuntimeException("Not support filter -> [like]");
-        }
-        if (funcDef.equals(BuiltInFunctionDefinitions.IN)) {
-            throw new RuntimeException("Not support filter -> [in]");
-        }
-        if (funcDef.equals(BuiltInFunctionDefinitions.BETWEEN)) {
-            throw new RuntimeException("Not support filter -> [between]");
-        }
+
         if (funcDef.equals(BuiltInFunctionDefinitions.CAST)) {
             return call.getChildren().get(0).accept(this);
         }
-        if (funcDef.equals(BuiltInFunctionDefinitions.NOT) && call.getOutputDataType().equals(DataTypes.BOOLEAN())) {
-            return call.getChildren().get(0).toString() + " = false";
-        }
-        if (funcDef.equals(BuiltInFunctionDefinitions.IS_NULL)) {
-            return call.getChildren().get(0).toString() + " is null";
-        }
-        if (funcDef.equals(BuiltInFunctionDefinitions.IS_NOT_NULL)) {
-            return call.getChildren().get(0).toString() + " is not null";
-        }
-        if (FUNC_TO_STR.containsKey(funcDef)) {
+
+        if (SUPPORT_FUNC.containsKey(funcDef)) {
             
             List<String> operands = new ArrayList<>();
             for (Expression child : call.getChildren()) {
                 String operand = child.accept(this);
                 if (operand == null) {
-                    continue;
+                    return null;
                 }
                 operands.add(operand);
             }
-            return "(" + String.join(" " + FUNC_TO_STR.get(funcDef) + " ", operands) + ")";
+            return "(" + SUPPORT_FUNC.get(funcDef).apply(operands.toArray(new String[0])) + ")";
         }
         return null;
     }

--- a/src/test/java/com/starrocks/connector/flink/table/source/StarRocksDynamicTableSourceTest.java
+++ b/src/test/java/com/starrocks/connector/flink/table/source/StarRocksDynamicTableSourceTest.java
@@ -14,30 +14,26 @@
 
 package com.starrocks.connector.flink.table.source;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-
-import java.util.Arrays;
-
 import com.starrocks.connector.flink.it.source.StarRocksSourceBaseTest;
 import com.starrocks.connector.flink.table.source.struct.PushDownHolder;
 
-import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.api.EnvironmentSettings;
 import org.apache.flink.table.api.TableEnvironment;
-import org.apache.flink.table.api.TableSchema;
-import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
-import org.apache.flink.table.api.internal.TableEnvironmentInternal;
-import org.apache.flink.table.delegation.Parser;
-import org.apache.flink.table.expressions.ResolvedExpression;
 import org.apache.flink.table.expressions.CallExpression;
 import org.apache.flink.table.expressions.FieldReferenceExpression;
+import org.apache.flink.table.expressions.ResolvedExpression;
 import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
 import static org.apache.flink.table.expressions.ApiExpressionUtils.valueLiteral;
+import static org.junit.Assert.assertEquals;
 
 public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
 
@@ -69,7 +65,18 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
     @Test
     public void testFilter() {
 
-        String filter = null;
+        String filter;
+
+        ResolvedExpression c9Ref = new FieldReferenceExpression("c6", DataTypes.STRING(), 0, 2);
+        ResolvedExpression c9CharLength = new CallExpression(BuiltInFunctionDefinitions.CHAR_LENGTH, Collections.singletonList(c9Ref), DataTypes.INT());
+        ResolvedExpression c9Exp =
+                new CallExpression(
+                        BuiltInFunctionDefinitions.LESS_THAN,
+                        Arrays.asList(c9CharLength, valueLiteral(10)),
+                        DataTypes.BOOLEAN());
+        dynamicTableSource.applyFilters(Collections.singletonList(c9Exp));
+        filter = pushDownHolder.getFilter();
+        Assert.assertTrue(filter.isEmpty());
 
         ResolvedExpression c5Ref = new FieldReferenceExpression("c5", DataTypes.TIMESTAMP(), 0, 2);
         ResolvedExpression c5Exp =
@@ -79,7 +86,7 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         DataTypes.BOOLEAN());
         dynamicTableSource.applyFilters(Arrays.asList(c5Exp));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("(c5 = '2022-1-22 00:00:00')"));
+        assertEquals("(c5 = '2022-1-22 00:00:00')", filter);
 
         ResolvedExpression c4Ref = new FieldReferenceExpression("c4", DataTypes.DATE(), 0, 2);
         ResolvedExpression c4Exp =
@@ -87,9 +94,9 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         BuiltInFunctionDefinitions.EQUALS,
                         Arrays.asList(c4Ref, valueLiteral("2022-1-22")),
                         DataTypes.BOOLEAN());
-        dynamicTableSource.applyFilters(Arrays.asList(c4Exp));
+        dynamicTableSource.applyFilters(Collections.singletonList(c4Exp));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("(c4 = '2022-1-22')"));
+        assertEquals("(c4 = '2022-1-22')", filter);
 
         ResolvedExpression c3Ref = new FieldReferenceExpression("c3", DataTypes.BOOLEAN(), 0, 2);
         ResolvedExpression c3Exp =
@@ -97,9 +104,9 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         BuiltInFunctionDefinitions.EQUALS,
                         Arrays.asList(c3Ref, valueLiteral(true)),
                         DataTypes.BOOLEAN());
-        dynamicTableSource.applyFilters(Arrays.asList(c3Exp));
+        dynamicTableSource.applyFilters(Collections.singletonList(c3Exp));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("(c3 = true)"));
+        assertEquals("(c3 = true)", filter);
 
         ResolvedExpression c2Ref = new FieldReferenceExpression("c2", DataTypes.INT(), 0, 2);
         ResolvedExpression c2Exp =
@@ -138,16 +145,16 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                 DataTypes.BOOLEAN())
         ));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("(c1 = 1) and (c1 <> 1) and (c1 > 1) and (c1 >= 1) and (c1 < 1) and (c1 <= 1)"));
+        assertEquals("(c1 = 1) and (c1 <> 1) and (c1 > 1) and (c1 >= 1) and (c1 < 1) and (c1 <= 1)", filter);
 
         dynamicTableSource.applyFilters(Arrays.asList(c1Exp, c2Exp));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("(c1 = 1) and (c2 = 2)"));
+        assertEquals("(c1 = 1) and (c2 = 2)", filter);
 
 
         dynamicTableSource.applyFilters(Arrays.asList(new CallExpression(BuiltInFunctionDefinitions.OR, Arrays.asList(c1Exp, c3Exp), DataTypes.BOOLEAN())));
         filter = pushDownHolder.getFilter();
-        assertTrue(filter.equals("((c1 = 1) or (c3 = true))"));
+        assertEquals("((c1 = 1) or (c3 = true))", filter);
 
 
         ResolvedExpression c6Exp =
@@ -156,10 +163,10 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         Arrays.asList(c1Ref, valueLiteral(1)),
                         DataTypes.BOOLEAN());
         try { 
-            dynamicTableSource.applyFilters(Arrays.asList(c6Exp));
+            dynamicTableSource.applyFilters(Collections.singletonList(c6Exp));
         } catch (Exception e) {
             e.printStackTrace();
-            assertTrue(e.getMessage().equals("Not support filter -> [like]"));
+            assertEquals("Not support filter -> [like]", e.getMessage());
         }
 
         ResolvedExpression c7Exp =
@@ -168,10 +175,10 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         Arrays.asList(c1Ref, valueLiteral(1)),
                         DataTypes.BOOLEAN());
         try { 
-            dynamicTableSource.applyFilters(Arrays.asList(c7Exp));
+            dynamicTableSource.applyFilters(Collections.singletonList(c7Exp));
         } catch (Exception e) {
             e.printStackTrace();
-            assertTrue(e.getMessage().equals("Not support filter -> [in]"));
+            assertEquals("Not support filter -> [in]", e.getMessage());
         }
 
         ResolvedExpression c8Exp =
@@ -180,10 +187,31 @@ public class StarRocksDynamicTableSourceTest extends StarRocksSourceBaseTest {
                         Arrays.asList(c1Ref, valueLiteral(1)),
                         DataTypes.BOOLEAN());
         try { 
-            dynamicTableSource.applyFilters(Arrays.asList(c8Exp));
+            dynamicTableSource.applyFilters(Collections.singletonList(c8Exp));
         } catch (Exception e) {
             e.printStackTrace();
-            assertTrue(e.getMessage().equals("Not support filter -> [between]"));
+            assertEquals("Not support filter -> [between]", e.getMessage());
         }
-    }   
+    }
+
+    @Test
+    public void test() {
+        EnvironmentSettings settings = EnvironmentSettings.newInstance().build();
+        TableEnvironment env = TableEnvironment.create(settings);
+//        env.explainSql("create table HTWSource (item_key string, vehicle_id string, item_type string, item_value string, modify_time timestamp) WITH ('connector' = 'datagen')");
+        env.executeSql("CREATE TABLE Orders (`user` BIGINT, product STRING, amount INT) WITH ('connector' = 'datagen')");
+        env.executeSql("CREATE TABLE HTWSource (`item_key` BIGINT, vehicle_id STRING, item_type STRING, item_value STRING, modify_time STRING) WITH ('connector' = 'datagen')");
+        System.out.println(env.explainSql(
+                "SELECT product, amount FROM Orders WHERE product LIKE '%Rubber%'"));
+
+        System.out.println(env.explainSql("select item_key,\n" +
+                "       vehicle_id,\n" +
+                "      item_type,item_value,modify_time\n" +
+                "from HTWSource \n" +
+                "where CHAR_LENGTH(vehicle_id) < 10"));
+
+        List<String> s = Collections.singletonList("a");
+
+        System.out.println(String.join(" is not null", s));
+    }
 }

